src/test/kotlin/org/sd_jwt/Utils.kt

```diff
@@ -58,7 +60,69 @@ inline fun <reified T>  testRoutine(
     assertEquals(expectedClaims, verifiedCredentialGen)
 }
 
+inline fun testRoutine(
+    expectedClaimsKeys: List<String>,
+    expectedClaims: JSONObject,
+    claims: JSONObject,
+    discloseStructure: JSONObject,
+    releaseClaims: JSONObject,
+    testConfig: TestConfig,
+    compareSingleValues: Boolean = false
+) {
+    println("\n====================================================")
+    println(testConfig.name)
+    println("====================================================\n")
+
+    // Initialization
+    val holderPubKey = testConfig.holderKey?.toPublicJWK()
+
+    val credentialGen = createCredential(
+        userClaims = claims,
+        issuerKey = testConfig.issuerKey,
+        holderPubKey = holderPubKey,
+        discloseStructure = discloseStructure
+    )
+
+    println("====================== Issuer ======================")
+    println("Generated credential: $credentialGen")
+
+    val presentationGen =
+        createPresentation(credentialGen, claims, releaseClaims, testConfig.verifier, testConfig.nonce, testConfig.holderKey)
 
+    println("====================== Wallet ======================")
+    println("Generated presentation: $presentationGen")
+
+    // Verify presentation
+    checkDisclosedDisclosures(presentationGen, expectedClaimsKeys)
+
+    // Raise an error if there is no holder binding, aud or nonce and the presentation does not end with a ~ character
+    if ((holderPubKey == null && testConfig.verifier == null && testConfig.nonce == null) && !presentationGen.endsWith("~")) {
+        throw Exception("Presentation without holder binding is missing '~' at the end")
+    }
+
+    val verifiedCredentialGen = verifyPresentation(
+        presentationGen, testConfig.trustedIssuers, testConfig.nonce, testConfig.verifier,
+        holderPubKey != null
+    )
+
+    println("===================== Verifier =====================")
+    println("Verified credential: $verifiedCredentialGen\n")
+
+    if(!compareSingleValues){
+        assertEquals(expectedClaims, verifiedCredentialGen)
+    }
+    else{
+        // verifies 2 (unsorted) JSONObject are matching
```

Comment:
Could we merge the two testRoutine functions and execute this check if the return type is a JSONObject? I would like to avoid code duplication.

Reply:
unfortunately this check is necessary since the structures to compare and not sorted and the full structures comparison will fail, thus comparing the json elements one by one
this check can be removed if we could sort the json objects the same way, but I found it simpler to do this check and not care how the json is presented.